zerver/tornado/handlers.py

```diff
@@ -339,16 +155,90 @@ def zulip_finish(self, result_dict: Dict[str, Any], request: HttpRequest,
+        #   could instead have just written the actual view code to
+        #   return the response we want, but that would result in
+        #   _get_response running Django's process_view middleware on
+        #   the HttpRequest a second time, which could result in weird
+        #   bugs).
+        #
+        # * By setting request.request_middleware_run, we configure
+        #   the below middleware_mixin__call___hack monkey-patching to
+        #   arrange the process_request middleware being skipped to
+        #   avoid running it twice, and instead only the
+        #   process_response middleware being run.
+        #
+        # * Then we just call `self.get_response`, passing our saved
+        #   HttpRequest, and send the resulting content to the client.
+
+        request.request_middleware_run = True
+        self.saved_response = json_response(res_type=result_dict['result'],
+                                            data=result_dict, status=self.get_status())
+        response = self.get_response(request)
+
         # Pass through the content-type from Django, as json content should be
         # served as application/json
-        self.set_header("Content-Type", django_response['Content-Type'])
-        self.finish(django_response.content)
+        #
+        # BUG: We should probably be copying all the headers by
+        # sharing code with the GET code above, rather than copying
+        # just this one header.
+        self.set_header("Content-Type", response['Content-Type'])
+
+        self.finish(response.content)
+
+    def _get_response(self, request: HttpRequest) -> HttpResponse:
+        # Small wrapper around the Django _get_response method to
+        # return an HttpResponse object stored in zulip_finish rather
+        # than the view function code a second time.  See zulip_finish
+        # for a detailed explanation of the design.
+        if self.saved_response is not None:
+            return self.saved_response
+        return super()._get_response(request)
+
+
+# Here we monkey-patch MiddlewareMixin to avoid calling the request
+# middleware twice when we call `get_response` a second time from a
+# long-poll request that we're waking up and having return in
+# zulip_finish.  The code is essentially the original implementation,
+# changed only to check `request_middleware_run`.  If this hack stops
+# working, it'll trigger the _log_data assert statement in the
+# LogRequests middleware's process_request method.
+def middleware_mixin__call___hack(self: MiddlewareMixin, request: HttpRequest) -> HttpResponse:
+    response = None  # type: Optional[HttpResponse]
+    # Zulip change: Added the request_middleware_run check on next line
+    if hasattr(self, 'process_request') and not getattr(request, "request_middleware_run", False):
+        response = self.process_request(request)
+    response = response or self.get_response(request)
+    if hasattr(self, 'process_response'):
```

Comment:
Would it make sense to skip the response middleware on the dummy response here with `and not getattr(response, "asynchronous", False)`?